### ISSUE TITLE
Compilation for android failing on windows - preprocessor sanity check fails

### ISSUE THREAD

**Issue by tomc1998** [alexcrichton/backtrace-rs#64]
When compiling a project that has backtrace-rs as a dependency, I get this:

```
"sh" "C:/Users/thoma/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-sys-0.1.12/src/libbacktrace/configure" "--with-pic" "--disable-multilib" "--disable-shared" "--disable-host-shared" "--host=arm-linux-androideabi"
running: "sh" "C:/Users/thoma/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-sys-0.1.12/src/libbacktrace/configure" "--with-pic" "--disable-multilib" "--disable-shared" "--disable-host-shared" "--host=arm-linux-androideabi"
configure: loading site script /usr/local/etc/config.site
checking build system type... x86_64-pc-msys
checking host system type... arm-unknown-linux-androideabi
checking target system type... arm-unknown-linux-androideabi
checking for arm-linux-androideabi-gcc... C:\Users\thoma\dev\tools\android-ndk-r14b\toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64\bin/arm-linux-androideabi-gcc
checking for C compiler default output file name... a.out
checking whether the C compiler works... yes
checking whether we are cross compiling... yes
checking for suffix of executables...
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether C:\Users\thoma\dev\tools\android-ndk-r14b\toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64\bin/arm-linux-androideabi-gcc accepts -g... yes
checking for C:\Users\thoma\dev\tools\android-ndk-r14b\toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64\bin/arm-linux-androideabi-gcc option to accept ISO C89... none needed
checking how to run the C preprocessor... /lib/cpp
```

After this, it fails. Here's stderr:

```
--- stderr
configure: WARNING: If you wanted to set the --build type, don't use --host.
    If a cross compiler is detected then cross compile mode will be used.
configure: error: in `/c/Users/thoma/dev/practice/rust-android/android_test/target/arm-linux-androideabi/debug/build/backtrace-sys-19334c94e1ccd578/out':
configure: error: C preprocessor "/lib/cpp" fails sanity check
See `config.log' for more details.
thread 'main' panicked at 'failed with: exit code: 1', C:\Users\thoma\.cargo\registry\src\github.com-1ecc6299db9ec823\backtrace-sys-0.1.12\build.rs:173
stack backtrace:
... stuff
             at .\build.rs:173
   7: build_script_build::main
             at .\build.rs:123
   8: _rust_maybe_catch_panic
... More stuff
```

Having a look at the config.log, I find this:

```
configure:3404: checking how to run the C preprocessor
configure:3435: C:\Users\thoma\dev\tools\android-ndk-r14b\toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64\bin/arm-linux-androideabi-gcc -E  conftest.c
In file included from c:\users\thoma\dev\tools\android-ndk-r14b\toolchains\arm-linux-androideabi-4.9\prebuilt\windows-x86_64\lib\gcc\arm-linux-androideabi\4.9.x\include-fixed\syslimits.h:7:0,
                 from c:\users\thoma\dev\tools\android-ndk-r14b\toolchains\arm-linux-androideabi-4.9\prebuilt\windows-x86_64\lib\gcc\arm-linux-androideabi\4.9.x\include-fixed\limits.h:34,
                 from conftest.c:10:
c:\users\thoma\dev\tools\android-ndk-r14b\toolchains\arm-linux-androideabi-4.9\prebuilt\windows-x86_64\lib\gcc\arm-linux-androideabi\4.9.x\include-fixed\limits.h:168:61: error: no include path in which to search for limits.h
 #include_next <limits.h>  /* recurse down to the real one */
                                                             ^
configure:3435: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "package-unused"
| #define PACKAGE_TARNAME "libbacktrace"
| #define PACKAGE_VERSION "version-unused"
| #define PACKAGE_STRING "package-unused version-unused"
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| /* end confdefs.h.  */
| #ifdef __STDC__
| # include <limits.h>
| #else
| # include <assert.h>
| #endif
| 		     Syntax error
```

And more following this, but with assert.h: No such file or directory.

I believe this might be because my environment is misconfigured, but I can compile a hello world program with GCC in C that includes assert.h. I figured it was an NDK problem (using NDK's gcc), but I've downloaded the latest version and the one before that, to no avail.

Is this an issue with including the --build flag somewhere in `configure`? I'd search through it, but it's like 10k lines long._. Do I just have a bad setup?

Thanks!

P.S.

Have tried this on both stable GNU and stable MSVC compilers on Windows 10.

**Comment by alexcrichton** [comment]
Hm are you sure that the "hello world" you're compiling is using the same compiler? Maybe backtrace-rs is using the wrong compiler by accident?

**Comment by tomc1998** [comment]
So I found out what the issue is - backtrace-rs IS using the correct compiler, but the sysroot is not set properly. When I was running my test hello world program, I mistyped the path (was using set path=<path to ndk gcc>) and it defaulted to my mingw install, silly me!

You need to set sysroot=<path to the lib files>, but that's different for every android arch you're compiling to. Say we're compiling to arm, android 18, we need: `--sysroot=$NDK_HOME/platforms/android-18/arch-arm`.

I don't really know what's the best way to go about adding these as flags, I'm far from a build.rs expert. In cargo-apk, which seems like pretty much the standard way to run rust on android, it uses custom variables in Cargo.toml. In here you can specify them as so:

```
[package.metadata.android]
# The target Android API level.
# It defaults to 18 because this is the minimum supported by rustc.
android_version = 18
# Specifies the array of targets to build for.
# Defaults to "arm-linux-androideabi".
# Other possible targets include "aarch64-linux-android", 
# "armv7-linux-androideabi", "i686-linux-android" and "x86_64-linux-android".
build_targets = [ "arm-linux-androideabi", "armv7-linux-androideabi" ]
```

As you can see, we could extract the android version from the .toml file and extract the arch path from the build target. But then you run the risk of limiting backtrace-rs compilation on android to the cargo-apk standard.

What would be the best way to go about specifying a sysroot for the android-ndk gcc?

**Comment by alexcrichton** [comment]
Hm I thought these compilers automatically found their sysroot? Or is this the case of a custom sysroot with a compiler located elsewhere?

**Comment by tomc1998** [comment]
Nope, the NDK compilers don't find the sysroot themselves. There's a build script called 'ndk-build' which takes a makefile - this sets the sysroot etc. This is for C projects though, and the plain GCC is used when compiling backtrace. See this stackoverflow answer for reference: https://stackoverflow.com/a/14308800

**Comment by siriux** [comment]
@tomc1998 are you using any workaround for this? I have the same problem, and I'm also using cargo-apk.

**Comment by tomc1998** [comment]
I never found a workaround. I think there was some clash between different versions of the compiler - unfortunately my windows PC has failed so I'm only on linux now, and the setup's a lot different! I figured some build script somewhere was using 'gcc' which got resolved to the system gcc as opposed to the android cross compiler.

Honestly it was so long ago that I couldn't give you any more details, but i'm pretty sure it WAS a config issue (and not an issue with backtrace-rs), i just couldn't fix it in the end / couldn't be sure enough to close the issue!

I'll close it now, since all this is probably in the wrong place. Good luck in your search<3
